lib/flags/local_flags.js

```diff
@@ -0,0 +1,368 @@
+            const rollout = flag.ruleset.rollout[index];
+
+            let salt;
+            if (flag.hash_salt !== null && flag.hash_salt !== undefined) {
+                salt = flag.key + flag.hash_salt + index.toString();
+            } else {
+                salt = flag.key + "rollout";
+            }
+
+            const rolloutHash = normalizedHash(String(contextValue), salt);
+
+            if (rolloutHash < rollout.rollout_percentage &&
+                this._isRuntimeEvaluationSatisfied(rollout, context)) {
+                return rollout;
+            }
+        }
+
+        return null;
+    }
+
+    _getAssignedVariant(flag, contextValue, flagKey, rollout) {
+        if (rollout.variant_override) {
+            const variant = this._getMatchingVariant(rollout.variant_override.key, flag);
+            if (variant) {
+                return { ...variant, is_qa_tester: false };
+            }
+        }
+
+        const storedSalt = flag.hash_salt !== null && flag.hash_salt !== undefined ? flag.hash_salt : "";
+        const salt = flagKey + storedSalt + "variant";
+        const variantHash = normalizedHash(String(contextValue), salt);
+
+        // Deep copy variants and apply splits
+        const variants = flag.ruleset.variants.map(v => ({ ...v }));
+        if (rollout.variant_splits) {
+            for (const variant of variants) {
+                if (variant.key in rollout.variant_splits) {
+                    variant.split = rollout.variant_splits[variant.key];
+                }
+            }
+        }
+
+        let selected = variants[0];
+        let cumulative = 0.0;
+        for (const variant of variants) {
+            selected = variant;
+            cumulative += variant.split || 0.0;
+            if (variantHash < cumulative) {
+                break;
+            }
+        }
+
+        return {
+            variant_key: selected.key,
+            variant_value: selected.value,
+            experiment_id: flag.experiment_id,
+            is_experiment_active: flag.is_experiment_active,
+            is_qa_tester: false,
+        };
+    }
+
+    _isRuntimeEvaluationSatisfied(rollout, context) {
+        if (!rollout.runtime_evaluation_definition) {
+            return true;
+        }
+
+        const customProperties = context.custom_properties;
+        if (!customProperties || typeof customProperties !== 'object') {
+            return false;
+        }
+
+        for (const [key, expectedValue] of Object.entries(rollout.runtime_evaluation_definition)) {
+            if (!(key in customProperties)) {
+                return false;
+            }
+
+            const actualValue = customProperties[key];
+            if (String(actualValue).toLowerCase() !== String(expectedValue).toLowerCase()) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    _trackExposure(flagKey, selectedVariant, context) {
```

Comment:
Looks to me like this is also quite similar between local+remote and could be extracted to the base class